File: src/main/resources/web/js/zkforge/canvas/Canvas.js
/* Canvas.js

	Purpose:
		Canvas prototype.
		
	Description:
		
	History:
		May 12, 2010 3:17:24 PM , Created by simon

Copyright (C) 2010 Potix Corporation. All Rights Reserved.

*/

/**
 * The ZK component corresponding to HTML 5 Canvas.
 * While HTML 5 Canvas is a command-based DOM object that allows user to draw
 * items on a surface, ZK Canvas maintains a list of drawable items and allow
 * user to operate the list by adding, removing, updating, replacing the
 * elements. The changes will be reflected on the client side upon these
 * operations.
 *
 * <p>Default {@link #getZclass}: z-canvas.
 */
zkforge.canvas.Canvas = zk.$extends(zul.Widget, {
    _cvs: null,
    _ctx: null,
    _drwbls: [],
    _states: [],

    // extended drawing states
    _drwTp: "fill",
    _drwTpBak: "fill",
    _txtMxW: -1,
    _txtMxWBak: -1,

    bind_: function () {
        this.$supers("bind_", arguments);

        this._cvs = document.createElement("canvas");

        var w = this.$n().clientWidth;
        var h = this.$n().clientHeight;
        if (w) this._cvs.width = zk.parseInt(w);
        if (h) this._cvs.height = zk.parseInt(h);

        jq(this.$n()).append(this._cvs);
        if (zk.ie) G_vmlCanvasManager.initElement(this._cvs);

        this._cvs.id = this.uuid + '-cnt';
        //if (!zk.ie) this._cvs.class = this.getZclass() + '-cnt'; // comfirm zclass spec

        this._ctx = this._cvs.getContext("2d");

    },

    setDrwngs: function (v) {
        this._drwbls = jq.evalJSON(v);
    },


    // add
    setAdd: function (drwJSON) {
        this.add(jq.evalJSON(drwJSON));
    },

    /**
     * Adds a Drawable to canvas.
     */
    add: function (drw) {
        this._paint(drw);
        this._drwbls.push(drw);
    },


    // add all
    setAddAll: function (drwsJSON) {
        this.addALL(jq.evalJSON(drwsJSON));
    },

    /**
     * Adds all the drawables in the array in the order.
     */
    addALL: function (drws) {
        var ln = drws.length;
        for (var i = 0; i < ln; i++) this._paint(drws[i]);
        this._drwbls = this._drwbls.concat(drws);
    },


    // remove
    setRemove: function (index) {
        this.remove(index);
    },

    /**
     * Removes the Drawable at specific index.
     */
    remove: function (index) {
        var drw = this._drwbls.splice(index, 1);
        this._repaint();
        return drw;
    },


    // remove all
    setRemoveAll: function (idsJSON) {
        this.removeAll(jq.evalJSON(idsJSON));
    },

    /**
     * Removes all the Drawables based on an array of indices
     */
    removeAll: function (indices) {
        for (var i = indices.length; i--;) this._drwbls.splice(indices[i], 1);
        this._repaint();
    },


    // retain all
    setRetainAll: function (idsJSON) {
        this.retainAll(jq.evalJSON(idsJSON));
    },

    /**
     * Retains all the Drawables based on an array of indices
     */
    retainAll: function (indices) {
        var drwbls2 = [];
        var ln = indices.length;
        for (var i = 0; i < ln; i++) drwbls2.push(this._drwbls[indices[i]]);
        this._drwbls = drwbls2;
        this._repaint();
    },


    // insert
    setInsert: function (idrwJSON) {
        var idrw = jq.evalJSON(idrwJSON);
        this.insert(idrw.i, idrw.drw);
    },

    /**
     * Inserts a Drawable at specific index.
     */
    insert: function (index, drw) {
        this._drwbls.splice(index, 0, drw);
        this._repaint();
    },


    // insert all
    setInsertAll: function (idrwsJSON) {
        var idrws = jq.evalJSON(idrwJSON);
        this.insertAll(idrws.i, idrws.drws);
    },

    /**
     * Inserts all Drawables in the list
     */
    insertAll: function (index, drws) {
        this._drwbls = this._drwbls.slice(0, index).concat(drws)
            .concat(this._drwbls.slice(index));
        this._repaint();
    },


    // replace
    setReplace: function (idrwJSON) {
        var idrw = jq.evalJSON(idrwJSON);
        this.replace(idrw.i, idrw.drw);
    },

    /**
     * Replace the Drawable at specific index.
     */
    replace: function (index, drw) {
        var removed = this._drwbls.splice(index, 1, drw);
        this._repaint();
        return removed[0];
    },


    // clear
    setClear: function () {
        this.clear();
    },

    /**
     * Remove all Drawables.
     */
    clear: function () {
        this._drwbls = [];
        this._clearCanvas();
    },


    // private //
    _clearCanvas: function () {
        this._ctx.clearRect(0, 0, this._cvs.width, this._cvs.height);
    },

    _repaint: function () {
        this._clearCanvas();
        var ln = this._drwbls.length;

        for (var i = 0; i < ln; i++) {
            this._paint(this._drwbls[i]);
        }
    },

    _paint: function (drw) {
        this._applyLocalState(drw.state);

        // the type value must match the return value of Drawable#getType()
        switch (drw.objtp) {
            case "rect":
                this._paintRect(drw.obj);
                break;
            case "path":
                this._paintPath(drw.obj);
                break;
            case "text":
                this._paintText(drw.obj);
                break;
            case "img":
            //break;
            default:
            // unsupported types
        }

        this._unapplyLocalState();
    },

    _paintRect: function (rect) {
        switch (this._drwTp) {
            case "none":
                break;
            case "stroke":
                this._ctx.strokeRect(rect.x, rect.y, rect.w, rect.h);
                break;
            case "both":
                // outline shadow
                this._ctx.save();
                this._ctx.globalAlpha = 0.01;
                this._ctx.strokeRect(rect.x, rect.y, rect.w, rect.h);
                this._ctx.restore();

                this._ctx.fillRect(rect.x, rect.y, rect.w, rect.h);

                // stroke outline without shadow
                this._ctx.save();
                this._clearShadow();
                this._ctx.strokeRect(rect.x, rect.y, rect.w, rect.h);
                this._ctx.restore();
                break;
            case "fill":
            default:
                this._ctx.fillRect(rect.x, rect.y, rect.w, rect.h);
        }
    },

    _paintPath: function (path) {
        // mimic path drawing based on data
        this._drawPath(path);

        switch (this._drwTp) {
            case "none":
                break;
            case "stroke":
                this._ctx.stroke();
                break;
            case "both":
                // outline shadow
                this._ctx.save();
                this._ctx.globalAlpha = 0.01;
                this._ctx.stroke();
                this._ctx.restore();

                this._ctx.fill();

                // stroke outline without shadow
                this._ctx.save();
                this._clearShadow();
                this._ctx.stroke();
                this._ctx.restore();
                break;
            case "fill":
            default:
                this._ctx.fill();
        }

        this._ctx.beginPath();
    },

    _drawPath: function (path) {
        var segments = path.sg;
        this._ctx.beginPath();

        for (var i = 0, len = segments.length; i < len; i++) {
            var data = segments[i].dt;
            switch (segments[i].tp) {
                case "mv":
                    this._ctx.moveTo(data[0], data[1]);
                    break;
                case "ln":
                    this._ctx.lineTo(data[0], data[1]);
                    break;
                case "qd":
                    this._ctx.quadraticCurveTo(data[0], data[1], data[2], data[3]);
                    break;
                case "bz":
                    this._ctx.bezierCurveTo(
                        data[0], data[1], data[2], data[3], data[4], data[5]);
                    break;
                // acrTo
                case "cl":
                    this._ctx.closePath();
            }
        }
    },

    _paintText: function (text) {
        switch (this._drwTp) {
            case "none":
                break;
            case "stroke":
                this._strkTxt(text);
                break;
            case "both":
                // outline shadow
                this._ctx.save();
                this._ctx.globalAlpha = 0.01;
                this._strkTxt(text);
                this._ctx.restore();

                this._filTxt(text);

                // stroke outline without shadow
                this._ctx.save();
                this._clearShadow();
                this._strkTxt(text);
                this._ctx.restore();
                break;
            case "fill":
            default:
                this._filTxt(text);
        }
    },

    _strkTxt: function (text) {
        if (this._txtMxW < 0) {
            this._ctx.strokeText(text.t, text.x, text.y);
        } else {
            this._ctx.strokeText(text.t, text.x, text.y, this._txtMxW);
        }
    },

    _filTxt: function (text) {
        if (this._txtMxW < 0) {
            this._ctx.fillText(text.t, text.x, text.y);
        } else {
            this._ctx.fillText(text.t, text.x, text.y, this._txtMxW);
        }
    },


    // state management helper //
    _applyLocalState: function (st) {
        // save current global state on DOM canvas context
        this._txtMxWBak = this._txtMxW;
        this._drwTpBak = this._drwTp;
        this._ctx.save();

        // apply local state to context
        this._setDOMContextState(st);
    },

    _setDOMContextState: function (st) {
        // this function applies the state of a drawable to DOM Canvas's global
        // state.

        if (st.dwtp) { // drawing type is NOT a part of DOM Canvas state
            this._drwTp = st.dwtp;
        }
        if (st.trns) { // transformation
            var trns = st.trns;
            this._ctx.setTransform(
                trns[0], trns[1], trns[2], trns[3], trns[4], trns[5]);
        }
        if (st.clp) { // clipping
            this._drawPath(st.clp.obj);
            this._ctx.clip();
        }

        if (st.strk) this._ctx.strokeStyle = st.strk;
        if (st.fil) this._ctx.fillStyle = st.fil;
        if (st.alfa) this._ctx.globalAlpha = st.alfa;
        if (st.lnw) this._ctx.lineWidth = st.lnw;
        if (st.lncp) this._ctx.lineCap = st.lncp;
        if (st.lnj) this._ctx.lineJoin = st.lnj;
        if (st.mtr) this._ctx.miterLimit = st.mtr;
        if (st.shx) this._ctx.shadowOffsetX = st.shx;
        if (st.shy) this._ctx.shadowOffsetY = st.shy;
        if (st.shb) this._ctx.shadowBlur = st.shb;
        if (st.shc) this._ctx.shadowColor = st.shc;
        if (st.cmp) this._ctx.globalCompositeOperation = st.cmp;
        if (st.fnt) this._ctx.font = st.fnt;
        if (st.txal) this._ctx.textAlign = st.txal;
        if (st.txbl) this._ctx.textBaseline = st.txbl;

        if (st.txmw) { // maxWidth is not a part of DOM Canvas state
            this._txtMxW = st.txmw;
        }
    },

    _unapplyLocalState: function () {
        // restore global state
        this._txtMxW = this._txtMxWBak;
        this._drwTp = this._drwTpBak;
        this._ctx.restore();
    },

    _clearShadow: function () {
        // remove shadow properties
        this._ctx.shadowOffsetX = null;
        this._ctx.shadowOffsetY = null;
        this._ctx.shadowBlur = null;
        this._ctx.shadowColor = null;
    },

    //super//
    getZclass: function () {
        var zcs = this._zclass;
        return zcs != null ? zcs : "z-canvas";
    }
});

File: src/main/java/com/gitlab/pedrioko/core/zk/component/Canvas.java
/* Canvas.java

{{IS_NOTE
	Purpose:
		Canvas prototype.
		
	Description:
		
	History:
		May 12, 2010 3:16:36 PM , Created by simon
}}IS_NOTE

Copyright (C) 2010 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
}}IS_RIGHT
*/
package com.gitlab.pedrioko.core.zk.component;

import com.gitlab.pedrioko.core.zk.component.model.Drawable;
import com.gitlab.pedrioko.core.zk.component.model.Path;
import com.gitlab.pedrioko.core.zk.component.model.Rectangle;
import com.gitlab.pedrioko.core.zk.component.model.Text;
import org.zkoss.json.JSONObject;
import org.zkoss.json.JSONValue;
import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.impl.XulElement;

import java.awt.geom.Rectangle2D;
import java.io.IOException;
import java.util.*;

/**
 * The prototype component corresponding to HTML 5 Canvas.
 * While HTML 5 Canvas is a command-based DOM object that allows user to draw
 * items on a surface, This Canvas maintains a list of drawable items and allow
 * user to operate the list by adding, removing, updating, replacing the
 * elements. The changes will be reflected on the client side upon these
 * operations.
 *
 * @author simon
 */
@SuppressWarnings("serial")
public class Canvas extends XulElement {
    private List<Drawable> _drawables;

    public Canvas() {
        _drawables = new LinkedList<Drawable>();
    }

    /* getter */

    /**
     * Return a list of all drawings in Canvas.
     */
    public List<Drawable> getAllDrawables() {
        return Collections.unmodifiableList(_drawables);
    }

    /**
     * Returns an iterator of Drawable list.
     */
    public Iterator<Drawable> iterator() {
        return _drawables.iterator();
    }

    /**
     * Returns a ListIterator of the Drawables
     * Returns a ListIterator of the Drawables
     */
    public ListIterator<Drawable> listIterator() {
        return _drawables.listIterator();
    }

    /**
     * Returns a ListIterator of the Drawables, starting at specific index.
     */
    public ListIterator<Drawable> listIterator(int index) {
        return _drawables.listIterator(index);
    }

    /**
     * Returns a view of the portion of this list between the specified
     * fromIndex, inclusive, and toIndex, exclusive.
     *
     * @param fromIndex
     * @param toIndex
     */
    public List<Drawable> subList(int fromIndex, int toIndex) {
        return _drawables.subList(fromIndex, toIndex);
    }

    /**
     * Returns an array containing all Drawable in the list.
     */
    public Drawable[] toArray() {
        return _drawables.toArray(new Drawable[0]);
    }

    /**
     * Returns the drawing at position index.
     *
     * @param index: drawings at 0 is the earliest drawing.
     */
    public Drawable get(int index) {
        return _drawables.get(index);
    }



    /* query */

    /**
     * Return true if the list is empty.
     */
    public boolean isEmpty() {
        return _drawables.isEmpty();
    }

    /**
     * Return true if contains the drawable.
     */
    public boolean contains(Drawable drawable) {
        return _drawables.contains(drawable);
    }

    /**
     * Return true if contains all the Drawables.
     *
     * @param drawables: A list of Drawables.
     */
    public boolean containsAll(Collection<Drawable> drawables) {
        return _drawables.containsAll(drawables);
    }

    /**
     * Returns the index of the first occurrence of the Drawable
     */
    public int indexOf(Drawable drawable) {
        return _drawables.indexOf(drawable);
    }

    /**
     * Returns the index of the last occurrence of the Drawable
     */
    public int lastIndexOf(Drawable drawable) {
        return _drawables.lastIndexOf(drawable);
    }

    /**
     * Returns the size of Drawable lists
     */
    public int size() {
        return _drawables.size();
    }



    /* operation */

    /**
     * Adds the Drawable object to the end of the list.
     *
     * @return true; See List API.
     */
    public boolean add(Drawable drawable) {
        smartUpdate("add", drawable.toJSONString(), true);
        return _drawables.add(drawable);
    }

    /**
     * Adds a Java 2D Shape.
     *
     * @param path: A Java 2D Shape
     * @return true; See List API.
     */
    public boolean add(java.awt.Shape shape) {
        if (shape instanceof java.awt.geom.Rectangle2D) {
            return add(new Rectangle((Rectangle2D) shape));
        } else {
            return add(new Path(shape));
        }
    }

    /**
     * Adds a Text. The reference point position depends on text align and text
     * baseline setting. See HTML 5 Canvas documentation.
     *
     * @param text: The String to be drawn.
     * @param x:    The X position.
     * @param y:    The Y position.
     * @return true; See List API.
     */
    public boolean add(String text, double x, double y) {
        return add(new Text(text, x, y));
    }

    /**
     * Inserts the Drawable at specific index
     */
    @SuppressWarnings("unchecked")
    public void add(int index, Drawable drawable) {
        _drawables.add(index, drawable);
        JSONObject args = new JSONObject();
        args.put("i", index);
        args.put("drw", drawable);
        smartUpdate("insert", args, true);
    }

    /**
     * Inserts a Shape at specific position
     *
     * @param shape: a Java 2D Shape
     */
    public void add(int index, java.awt.Shape shape) {
        if (shape instanceof java.awt.geom.Rectangle2D) {
            add(index, new Rectangle((Rectangle2D) shape));
        } else {
            add(index, new Path(shape));
        }
    }

    /**
     * Inserts a Text at specific index. The reference point position depends
     * on text align and text baseline setting. See HTML 5 Canvas documentation.
     *
     * @param index: The position in the list to insert Text.
     * @param text:  The String to be drawn.
     * @param x:     The X position.
     * @param y:     The Y position.
     */
    public void add(int index, String text, double x, double y) {
        add(index, new Text(text, x, y));
    }

    /**
     * Adds all drawables in the list, in the given order.
     *
     * @param drawables: A List of Drawable
     * @return true; See List API.
     */
    public boolean addAll(List<Drawable> drawables) {
        smartUpdate("addAll", JSONValue.toJSONString(drawables), true);
        return _drawables.addAll(drawables);
    }

    /**
     * Inserts all drawables in the list at the specific index
     *
     * @param index:     the index to insert the Drawables
     * @param drawables: A List of Drawable
     * @return true if the operation changes the Canvas Drawable list
     */
    @SuppressWarnings("unchecked")
    public boolean addAll(int index, List<Drawable> drawables) {
        JSONObject args = new JSONObject();
        args.put("i", index);
        args.put("drws", drawables);
        smartUpdate("insertAll", args, true);
        return _drawables.addAll(index, drawables);
    }

    /**
     * Removes the Drawable at specific index.
     *
     * @return The removed Drawable
     */
    public Drawable remove(int index) {
        smartUpdate("remove", index, true);
        return _drawables.remove(index);
    }

    /**
     * Removes a Drawable.
     *
     * @return: true if the list is changed as a result.
     */
    public boolean remove(Drawable drawable) {
        smartUpdate("remove", _drawables.indexOf(drawable), true);
        return _drawables.remove(drawable);
    }

    /**
     * Removes all Drawables contained in the list
     *
     * @return: true if the list is changed as a result.
     */
    public boolean removeAll(Collection<Drawable> drawables) {
        List<Integer> indices = new ArrayList<Integer>();
        for (int i = 0; i < _drawables.size(); i++) {
            if (drawables.contains(_drawables.get(i))) indices.add(i);
        }
        smartUpdate("removeAll", JSONValue.toJSONString(indices));
        return _drawables.removeAll(drawables);
    }

    /**
     * Retains all Drawables contained in the list
     *
     * @return: true if the list is changed as a result.
     */
    public boolean retainAll(Collection<Drawable> drawables) {
        List<Integer> indices = new ArrayList<Integer>();
        for (int i = 0; i < _drawables.size(); i++) {
            if (drawables.contains(_drawables.get(i))) indices.add(i);
        }
        smartUpdate("retainAll", JSONValue.toJSONString(indices));
        return _drawables.retainAll(drawables);
    }


    /**
     * Clears the Drawable list. The Canvas is also cleared as a result.
     */
    public void clear() {
        _drawables.clear();
        smartUpdate("clear", null);
    }

    /**
     * Replace a Drawable at specific index.
     *
     * @return The replaced Drawable
     */
    @SuppressWarnings("unchecked")
    public Drawable set(int index, Drawable drawable) {
        JSONObject args = new JSONObject();
        args.put("i", index);
        args.put("drw", drawable);
        smartUpdate("replace", args, true);
        return _drawables.set(index, drawable);
    }

    /**
     * Replace a Shape at specific index.
     *
     * @return The replaced Drawable
     */
    public Drawable set(int index, java.awt.Shape shape) {
        if (shape instanceof java.awt.geom.Rectangle2D) {
            return set(index, new Rectangle((Rectangle2D) shape));
        } else {
            return set(index, new Path(shape));
        }
    }

    /**
     * Replaces a Text at specific index. The reference point position depends
     * on text align and text baseline setting. See HTML 5 Canvas documentation.
     *
     * @param index: The index of Drawable to be replaced.
     * @param text:  The String to be drawn.
     * @param x:     The X position.
     * @param y:     The Y position.
     * @return The replaced Drawable
     */
    public Drawable set(int index, String text, double x, double y) {
        return set(index, new Text(text, x, y));
    }


    /* super */
    @Override
    protected void renderProperties(ContentRenderer renderer)
            throws IOException {

        super.renderProperties(renderer);
        render(renderer, "drwngs", JSONValue.toJSONString(_drawables));
    }

    @Override
    public String getZclass() {
        return _zclass == null ? "z-canvas" : _zclass;
    }

}
